Handling of unsigned values

rjo12 opened the pull request:
**Describe the pull request**

(1) Correction of bug related to calculating difference of "unsigned" values.   Essentially an "off by one" error.     If an unsigned variable has range 0 to UINT_MAX,  numeric operations on those variables are modulo UINT_MAX + 1 (e.g. subtracting 1 from 0 produces UINT_MAX).    The bug was doing operations modulo UINT_MAX, not modulo UINT_MAX + 1 when subtracting a large value from a small one.

(2)   Correction of test case that allowed preceding bug (1) to pass!

(3)    When calculating UintUtils.diff(a, b, UINT_MAX), no longer throw exceptions if a or b are out of range [0, UINT_MAX].   Instead, correct the values a and b using modulo arithmetic, and then calculate their difference.   This is "normal" handling of numeric operations unsigned integral types, both in hardware and in other programming languages. 

(4)    Add test cases to verify behaviour of code added in (3)

(5)   Existing test cases were only for "16 bit" unsigned subtraction.   "32 bit" tests added in interests of completeness (verify rather than assume that passing 16-bit tests mean that 32-bit tests will pass).

Edits in this pull request MAY address some contributors to this and other bugs related to spikes in sensor data (e.g. my experience of a recorded speed of 2394.0 km/hr on a bicycle, despite belief I did not achieve such a speed :-) ).    The nature of the bug is that it can map a small value to a large value, and vice versa.

**Link to the issue**

Bug (1 above) with handling calculation of difference of unsigned values is mentioned in https://github.com/OpenTracksApp/OpenTracks/issues/1557

**License agreement**
By opening this pull request, you are providing your contribution under the _Apache License 2.0_ (see [LICENSE.md](LICENSE.md)).

Acknowledged.

**Note: new dependencies/libraries**
Please refrain from introducing new libraries without consulting the team.

Refrained!

dennisguse commented on src/main/java/de/dennisguse/opentracks/sensors/UintUtils.java:
```diff
@@ -14,17 +14,20 @@ private UintUtils() {
-        if (a < 0 || b < 0) {
-            throw new RuntimeException("a or b cannot be less than zero.");
+        
+        a %= (UINT_MAX + 1);
```
Crashing on overflow is intended.
This will show if something else in the whole Bluetooth parsing is broken: crash early.
Otherwise a problem might be hidden for ages.

rjo12 commented:
OK.   I've added the "crash on overflow" back in, and modified test cases accordingly (wrapping tests of values out of range in an "if (false)" block, since I suggest they are appropriate in future).

I haven't removed the corrections to values of a and b, since they have no effect if a and b are in range (effectively a no-op). 

I've done it that way because, in the long run, I suggest it would be better to catch parsing errors by constructing test cases for the parsing code, and not crashing out of sensor-related code.  [Although I accept that would be a challenging job,  I also suggest it's not really the role of sensors to detect parsing errors].

Crashing out here also means crashing out if the parser is working correctly (i.e. extracting a value that the sensor sent) but has been sent data out of range by a buggy sensor.    In that case, the appropriate action is to correct a and b (which is what the sensor should have done) and not crash out.   Handling a buggy sensor is a job appropriate for sensor-related code rather than the parser, and would also avoid inflicting crashes "in the wild"  on users unlucky enough to have bought such a sensor when we can easily deal with the problem.

dennisguse commented:
One final thing: can you squash the commits and set a useful commit message.
Something like: 
`Bugfix: UINT parsing overflow of by one.
Fixes #1557`

And can you remove the code handling the overflow?
Keeping the test cases is fine!

PS/ I would love to push a new release tomorrow and this fix must be definitely in it!

rjo12 commented:
OK;  done!

I've started looking at the code that is extracting the values from bluetooth for each sensor.   Although not for today, I am concerned that doing the overflow check in the diff function and crashing out there may be too late in some cases  (essentially, an overflowed value is only picked up if calculating its difference from another value).

dennisguse commented:
One final thing: can you squash the commits into one (and also drop the merge commit)?

rjo12 commented:
I'm at a loss.

I've been doing all this through a browser on github.com.     I've searched, and can't find any mechanism to squash the commits, or drop the merge commit (which I hadn't realised I had done).

Instructions I find to squash commits on docs.github.com describe a process using Github Desktop, but I haven't managed to find anything even remotely comparable via the github.com site.

I'm probably missing something blindingly obvious, but I've just lost a couple of hours going round in circles trying to work out how.

dennisguse commented:
Then leave it as it is.

For squashing you have to use a local checkout if the repo: https://www.baeldung.com/ops/git-squash-commits

We can also squash while merging via GitHub (just not that cool).

dennisguse commented:
Manually merged: 7208d2a346604a96f5ac68f33674517f16960b2b
